zallet/src/commands/migrate_zcashd_wallet.rs

```diff
@@ -0,0 +1,664 @@
+        // * The mnemonic HD seed, under a standard ZIP 32 key path
+        // * The mnemonic HD seed, under the "legacy" account with an additional hardened path element
+        // * Zcashd Sapling spending key import
+        for (idx, key) in wallet.sapling_keys().keypairs().enumerate() {
+            // `zewif_zcashd` parses to an earlier version of the `sapling` types, so we
+            // must roundtrip through the byte representation into the version we need.
+            let extsk = sapling::zip32::ExtendedSpendingKey::from_bytes(&key.extsk().to_bytes())
+                .map_err(|_| ()) //work around missing Debug impl
+                .expect("Sapling extsk encoding is stable across sapling-crypto versions");
+            #[allow(deprecated)]
+            let extfvk = extsk.to_extended_full_viewing_key();
+            let ufvk =
+                UnifiedFullViewingKey::from_sapling_extended_full_viewing_key(extfvk.clone())?;
+
+            let key_seed_fp = key
+                .metadata()
+                .seed_fp()
+                .map(|seed_fp_bytes| SeedFingerprint::from_bytes(*seed_fp_bytes.as_bytes()));
+
+            let derivation = key
+                .metadata()
+                .hd_keypath()
+                .map(|keypath| ZcashdHdDerivation::parse_hd_path(&network_params, keypath))
+                .transpose()?
+                .zip(key_seed_fp)
+                .map(|(derivation, key_seed_fp)| match derivation {
+                    ZcashdHdDerivation::Zip32 { account_id } => {
+                        Zip32Derivation::new(key_seed_fp, account_id, None)
+                    }
+                    ZcashdHdDerivation::Post470LegacySapling { address_index } => {
+                        Zip32Derivation::new(
+                            key_seed_fp,
+                            ZCASHD_LEGACY_ACCOUNT,
+                            Some(address_index),
+                        )
+                    }
+                });
+
+            // If the key is not associated with either of the seeds, treat it as a standalone
+            // imported key
+            if key_seed_fp != mnemonic_seed_fp && key_seed_fp != legacy_seed_fp {
+                keystore
+                    .encrypt_and_store_standalone_sapling_key(&extsk)
+                    .await?;
+            }
+
+            let account_exists = match key_seed_fp.as_ref() {
+                Some(fp) => db_data
+                    .get_derived_account(&Zip32Derivation::new(
+                        *fp,
+                        ZCASHD_LEGACY_ACCOUNT,
+                        derivation.as_ref().and_then(|d| d.legacy_address_index()),
+                    ))?
+                    .is_some(),
+                None => db_data.get_account_for_ufvk(&ufvk)?.is_some(),
+            };
+
+            if !account_exists {
+                db_data.import_account_ufvk(
+                    &format!("zcashd legacy sapling {}", idx),
```

Comment:
Non-blocking, but for the record this will be incorrect for imported Sapling keys (which are not associated with any legacy account) until #263 is done.